aiidalab/app.py

```diff
@@ -82,14 +82,131 @@ class AiidaLabApp(traitlets.HasTraits):
     busy = traitlets.Bool(readonly=True)
     modified = traitlets.Bool(readonly=True, allow_none=True)
 
+    class _GitReleaseLine:
+        """Utility class to operate on the release line of the app.
+
+        A release line is specified via the app url as the part behind the '@'.
+
+        A release line can be specified either as
+            a) a commit denoted by a hexadecimal number with either 20 or 40 digits, or
+            b) a short reference, which can be either a branch or a tag name.
+
+        A full ref is the ref as defined in the Git glossary, e.g., 'refs/heads/main'.
+        A full ref and a commit are a revision.
+        """
+
+        def __init__(self, app, line):
+            self.app = app
+            self.line = line
+
+            match = re.fullmatch(r'(?P<commit>([0-9a-fA-F]{20}){1,2})|(?P<short_ref>.+)', line)
+            if not match:
+                raise ValueError(f"Illegal release line: {line}")
+
+            self.commit = match.groupdict()['commit']
+            self.short_ref = match.groupdict()['short_ref']
+            assert self.commit or self.short_ref
+
+        @property
+        def _repo(self):
+            return Repo(self.app.path)
+
+        def _resolve_short_ref(self, short_ref):
+            """Attempt to resolve the short-ref to a full ref.
+
+            For example, 'branch' would be resolved to 'refs/heads/branch'
+            if 'branch' is a local branch or 'refs/tags/branch' if it was
+            a tag.
+
+            This function returns None if the short-ref cannot be resolved
+            to a full reference.
+            """
+            # Check if short-ref is a head:
+            if f'refs/heads/{short_ref}'.encode() in self._repo.refs.allkeys():
+                return f'refs/heads/{short_ref}'.encode()
+
+            # Check if short-ref is a tag:
+            if f'refs/tags/{short_ref}'.encode() in self._repo.refs.allkeys():
+                return f'refs/tags/{short_ref}'.encode()
+
+            # Check if short-ref is among the remote refs:
+            for ref in self._repo.refs.allkeys():
+                if re.match(r'refs\/remotes\/(.*)?\/' + short_ref, ref.decode()):
+                    return ref
+
+            return None
+
+        def find_versions(self):
+            """Find versions available for this release line.
+
+            When encountering an ambiguous release line name, i.e.,
+            a shared branch and tag name, we give preference to the
+            branch, because that is what git does in this situation.
+            """
+            assert self.short_ref or self.commit
+
+            if self.commit:  # The release line is a commit.
+                assert self.commit.encode() in self._repo.object_store
+                yield self.commit.encode()
+            else:
+                ref = self._resolve_short_ref(self.short_ref)
+                if ref is None:
+                    raise ValueError(f"Unable to resolve '{self.short_ref}'. "
```

Comment:
Casting as `repr()` does the same. A nice trick :)
```suggestion
                    raise ValueError(f"Unable to resolve {self.short_ref!r}. "
```